### PR TITLE
FailFast - configuration

### LINKED ISSUE #2480: Configuration to control fail fast behavior during application startup
# Feature Request

**Are you requesting automatic instrumentation for a framework or library? Please describe.**
No

**Is your feature request related to a problem? Please describe.**
The Rules Engine is useful for detecting issue that we are aware of. However, there may be some edge cases that we are not aware of that could cause a crash during application startup. 

**Describe the solution you'd like**
We should provide a configuration that allows the application to handle errors more gracefully during application startup and prevent the application from crashing. For example, it may be useful to fail quickly and crash in a dev or test environment, but in production we do not want to cause the application to crash.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by Kielek**
## Why

Fixes #2480

## What

Support for Fail Fast while parsing data.

## Tests

CI + newly created tests.

## Checklist

- ~~[ ] `CHANGELOG.md` is updated.~~
- ~~[ ] Documentation is updated.~~
- [x] New features are covered by tests.

### CODE REVIEW

**Review by RassK — APPROVED** [review]

File: src/OpenTelemetry.AutoInstrumentation/Configurations/Settings.cs [hunk 1]
```diff
@@ -28,10 +28,10 @@ internal abstract class Settings
     /// </summary>
     public OtlpExportProtocol? OtlpExportProtocol { get; private set; }
 
-    public static T FromDefaultSources<T>()
+    public static T FromDefaultSources<T>(bool failFast)
         where T : Settings, new()
     {
-        var configuration = new Configuration(new EnvironmentConfigurationSource());
+        var configuration = new Configuration(failFast, new EnvironmentConfigurationSource(failFast));
```
Comment (zacharycmontoya) [inline src/OpenTelemetry.AutoInstrumentation/Configurations/Settings.cs:34]: This seems like a chicken and egg problem, trying to embed a `failFast` value into the configuration source even though its value can be provided from a configuration source. I'm not requesting any changes, but this looks like an interesting circular dependency

File: src/OpenTelemetry.AutoInstrumentation/Instrumentation.cs [hunk 2]
```diff
@@ -242,6 +244,11 @@ private static void AddLazilyLoadedMetricInstrumentations(LazyInstrumentationLoa
                     break;
                 default:
                     Logger.Warning($"Configured metric instrumentation type is not supported: {instrumentation}");
+                    if (FailFastSettings.Value.FailFast)
```
Comment (zacharycmontoya) [inline src/OpenTelemetry.AutoInstrumentation/Instrumentation.cs:247]: Will this be overridden if the FAIL FAST environment variable is set? Is there already a smoke test that covers this scenario?

**Review by zacharycmontoya — APPROVED** [review]

**Review by nrcventura — APPROVED** [review]

File: src/OpenTelemetry.AutoInstrumentation/Configurations/TracerSettings.cs [hunk 3]
```diff
@@ -63,7 +63,7 @@ internal class TracerSettings : Settings
     /// <summary>
     /// Gets the instrumentation options.
     /// </summary>
-    public InstrumentationOptions InstrumentationOptions { get; private set; } = new(new Configuration());
+    public InstrumentationOptions InstrumentationOptions { get; private set; } = new(new Configuration(false));
```
Comment (pjanotti) [inline src/OpenTelemetry.AutoInstrumentation/Configurations/TracerSettings.cs:66]: Is it possible to use a named parameter consistently when creating types that take it as a parameter?

File: src/OpenTelemetry.AutoInstrumentation/Configurations/FailFastSettings.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace OpenTelemetry.AutoInstrumentation.Configurations;
+
+internal class FailFastSettings : Settings
+{
+    public bool FailFast { get; private set; }
+
+    protected override void OnLoad(Configuration configuration)
+    {
+        FailFast = configuration.GetBool(ConfigurationKeys.FailFast) ?? false;
```
Comment (pjanotti) [inline src/OpenTelemetry.AutoInstrumentation/Configurations/FailFastSettings.cs:25]: 😄 can we make fail fast fail fast if fail fast fail?

File: test/OpenTelemetry.AutoInstrumentation.Tests/Configurations/ConfigurationTests.cs [hunk 5]
```diff
@@ -35,7 +35,7 @@ private enum TestEnum
     [Fact]
     public void ParseEnabledEnumList_Default_Enabled()
     {
-        var source = new Configuration(new NameValueConfigurationSource(new NameValueCollection()));
+        var source = new Configuration(false, new NameValueConfigurationSource(false, new NameValueCollection()));
```
Comment (pjanotti) [inline test/OpenTelemetry.AutoInstrumentation.Tests/Configurations/ConfigurationTests.cs:38]: For the tests, I don't think that naming the parameter is worth it.

**Review by pjanotti — APPROVED** [review]

File: src/OpenTelemetry.AutoInstrumentation/Configurations/TracerSettings.cs (hunk 3 above)
Comment (Kielek) [reply src/OpenTelemetry.AutoInstrumentation/Configurations/TracerSettings.cs:66]: It was the only place under src with non-named parameter. Fixed.

File: src/OpenTelemetry.AutoInstrumentation/Configurations/FailFastSettings.cs (hunk 4 above)
Comment (Kielek) [reply src/OpenTelemetry.AutoInstrumentation/Configurations/FailFastSettings.cs:25]: :D no!

**Pull request merged**